[LC-1008] Edit logic to build block without `rootHash` after `invoke`.

hyerim-kim opened the pull request:
Edit logic to build block without `rootHash` after `invoke`.

- Remove some duplicated code.
- Add some functions into BlockBuilder of version 0.3.

hyerim-kim committed:
[LC-1008] Edit logic to build block without `rootHash` after `invoke`.

- Remove some duplicated code.
- Add some functions into BlockBuilder of version 0.3.

metalg0su commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1176,19 +1175,19 @@ def _process_next_prep_legacy(self, _block: Block, block_builder: BlockBuilder,
-    def _process_next_prep(self, _block: Block, block_builder: BlockBuilder, next_prep: dict):
-        reps = self.find_preps_addresses_by_header(_block.header)
-
+    def _process_next_prep(self, _block: Block, block_builder: BlockBuilder, reps: list, next_prep: dict):
```
> I'm afraid that this review is out of the PR's topic...

param `next_prep` could be `None`

I suggest:
```python
def _process_next_prep(..., next_prep: Union[dict]):
...
```

OR

```python
next_prep = response.get("prep", {}) # at 1285 line
```

metalg0su commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1164,7 +1163,7 @@ def genesis_invoke(self, block: Block, prev_block_=None) -> Tuple[Block, dict]:
         return new_block, tx_receipts
 
-    def _process_next_prep_legacy(self, _block: Block, block_builder: BlockBuilder, next_prep: dict):
+    def _process_next_prep_legacy(self, _block: Block, block_builder: BlockBuilder, reps: list, next_prep: dict):
```
See:

metalg0su commented on loopchain/blockchain/blocks/v0_3/block_builder.py:
```diff
@@ -129,20 +130,27 @@ def _build_receipts_hash(self):
+        return self.next_reps_hash
+
+    def _build_next_reps_hash(self):
+        block_prover = BlockProver((rep.extend() for rep in self.next_reps), BlockProverType.Rep)
```
How about checking `self.reps == self.next_reps` first, then use `self.reps_hash` instead, if two reps are identical?

I'm not sure but it might be faster in certain situations.

> get root hash of all leaves ALWAYS VS compare all items of sequence

hyerim-kim replied:
I almost agree with you, most cases are the same maybe. But `reps` must be built earlier than `next_reps` always if do as your opinion. I worry about the dependency of the sequence between them. Let me know your opinion.

metalg0su replied:
You're right.

My suggestion above is assured IF the reps hash is built BEFORE next reps hash.

Please keep it on.

hyerim-kim committed:
[LC-1008] Add feedback.

- Prevent empty data.
- Figure `next_reps_hash` out directly instead of referencing from a `block head` in the function of `__write_block_data()`.
- Set alias to avoid confusing with the superclass that has same name.
- Edit align of some code.

windies21 commented on loopchain/blockchain/blockchain.py:
```diff
@@ -591,8 +591,7 @@ def __add_block(self, block: Block, confirm_info, need_to_write_tx_info=True, ne
 
-            if next_prep and self.find_preps_addresses_by_roothash(
-                    Hash32.fromhex(next_prep['rootHash'], ignore_prefix=True)):
+            if next_prep and self.find_preps_addresses_by_roothash(block.header.next_reps_hash):
```
For block version 0.3. next_reps_hash is not always filled. See block specifications.  If you want the actual value of next_reps_hash, use reveal_next_reps_hash.

windies21 replied:
I think the way to compute the hash from next_prep itself is safer than using block's next_reps_hash.

windies21 commented on loopchain/blockchain/blockchain.py:
```diff
@@ -694,10 +693,10 @@ def __write_block_data(self, block: Block, confirm_info, receipts, next_prep):
-                f"store next_prep in __write_block_data\nprep_hash({next_prep['rootHash']})"
-                f"\npreps({next_prep['preps']})")
-            self.write_preps(Hash32.fromhex(next_prep['rootHash'], ignore_prefix=True), next_prep['preps'], batch)
+            utils.logger.spam(f"store next_prep in __write_block_data\n"
```
My suggestion is calculate and use the reps_hash in this method.

windies21 commented on loopchain/blockchain/blocks/v0_4/block_builder.py:
```diff
@@ -1,12 +1,8 @@
-
-from loopchain.blockchain.blocks import BlockBuilder as BaseBlockBuilder, BlockProverType
 from loopchain.blockchain.blocks.block import NextRepsChangeReason
 from loopchain.blockchain.blocks.v0_3 import BlockBuilder
```
How about this:
from loopchain.blockchain.blocks.v0_3 import BlockBuilder as BaseBlockBuilder

windies21 commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1176,19 +1169,19 @@ def _process_next_prep_legacy(self, _block: Block, block_builder: BlockBuilder,
         else:
             # P-Rep list has no changes
-            next_leader = _block.header.next_leader
+            next_preps = reps
```
Does it fit the v0.3 block specification? see. https://icon-project.atlassian.net/wiki/spaces/LO/pages/92045385/Block+0.3
